store/store_download.go

```diff
@@ -646,16 +653,38 @@ var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo,
 	partialTargetPath := targetPath + ".partial"
 
 	xdelta3Args := []string{"-d", "-s", snapPath, deltaPath, partialTargetPath}
-	cmd, err := getXdelta3Cmd(xdelta3Args...)
+	cmd, isSystemSnapProvided, err := getXdelta3Cmd(xdelta3Args...)
 	if err != nil {
 		return err
 	}
 
-	if err := cmd.Run(); err != nil {
-		if err := os.Remove(partialTargetPath); err != nil {
-			logger.Noticef("failed to remove partial delta target %q: %s", partialTargetPath, err)
+	// try to run the xdelta command, optionally falling back and cleaning up
+	// afterwards
+	if runErr := cmd.Run(); runErr != nil {
```

Comment:
Maybe we can use a slightly more linear way of doing this? Not sure if that is possible but I sketched an idea below.

 Maybe something similar to:
```go
diff --git a/store/store_download.go b/store/store_download.go
index c7095c8439..6b85cae9e5 100644
--- a/store/store_download.go
+++ b/store/store_download.go
@@ -83,8 +83,11 @@ func init() {
diff --git a/store/store_download.go b/store/store_download.go
index c7095c8439..ae4f189d10 100644
--- a/store/store_download.go
+++ b/store/store_download.go
@@ -83,7 +83,7 @@ func init() {
 func useDeltas() bool {
        // only xdelta3 is supported for now, so check the binary exists here
        // TODO: have a per-format checker instead
-       if _, _, err := getXdelta3Cmd(); err != nil {
+       if _, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3"); err != nil && !osutil.ExecutableExists("xdelta3") {
                return false
        }
 
@@ -598,7 +601,7 @@ func doDownloadReqImpl(ctx context.Context, storeURL *url.URL, cdnHeader string,
 }
 
 // downloadDelta downloads the delta for the preferred format, returning the path.
-func (s *Store) downloadDelta(deltaName string, downloadInfo *snap.DownloadInfo, w io.ReadWriteSeeker, pbar progress.Meter, user *auth.UserState, dlOpts *DownloadOptions) error {
+func (s *Store) downloadDelta(deltaName string, downloadInfo *snap.DownloadInfo, w io.ReadWriteSeeker, pbar progress.Meter, user *auth.UserState, dlOpts *DownloadOptions) (err error) {
 
        if len(downloadInfo.Deltas) != 1 {
                return errors.New("store returned more than one download delta")
@@ -623,22 +626,8 @@ func (s *Store) downloadDelta(deltaName string, downloadInfo *snap.DownloadInfo,
        return download(context.TODO(), deltaName, deltaInfo.Sha3_384, url, user, s, w, 0, pbar, dlOpts)
 }
 
-func getXdelta3Cmd(args ...string) (*exec.Cmd, bool, error) {
-       cmd, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3", args...)
-       if err != nil {
-               // try checking the system provided version
-               if osutil.ExecutableExists("xdelta3") {
-                       return exec.Command("xdelta3", args...), false, nil
-               }
-
-               return nil, false, fmt.Errorf("no xdelta3 available")
-       }
-
-       return cmd, true, nil
-}
-
 // applyDelta generates a target snap from a previously downloaded snap and a downloaded delta.
-var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo, targetPath string, targetSha3_384 string) error {
+var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo, targetPath string, targetSha3_384 string) (err error) {
        snapBase := fmt.Sprintf("%s_%d.snap", name, deltaInfo.FromRevision)
        snapPath := filepath.Join(dirs.SnapBlobDir, snapBase)
 
@@ -651,39 +640,26 @@ var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo,
        }
 
        partialTargetPath := targetPath + ".partial"
-
-       xdelta3Args := []string{"-d", "-s", snapPath, deltaPath, partialTargetPath}
-       cmd, isSystemSnapProvided, err := getXdelta3Cmd(xdelta3Args...)
-       if err != nil {
-               return err
-       }
-
-       // try to run the xdelta command, optionally falling back and cleaning up
-       // afterwards
-       if runErr := cmd.Run(); runErr != nil {
-               if isSystemSnapProvided {
-                       // fallback to trying the version on $PATH instead
-                       if !osutil.ExecutableExists("xdelta3") {
-                               // internal error, we should have already checked if deltas are
-                               // available before calling this function
-                               return fmt.Errorf("internal: no xdelta3 available")
-                       }
-                       pathCmd := exec.Command("xdelta3", xdelta3Args...)
-                       if runErr := pathCmd.Run(); runErr != nil {
-                               if err := os.Remove(partialTargetPath); err != nil {
-                                       logger.Noticef("after encountering error applying delta: %v, also failed to remove partial delta target %q: %s", runErr, partialTargetPath, err)
-                               }
-                       }
-                       // otherwise we succeeded with the $PATH version, but still failed
-                       // with the system provided one, so log that error in case we end up
-                       // having issues there
-                       logger.Noticef("overcame error from %q from system snap (%v) by using $PATH version (%q)", cmd.Path, runErr, pathCmd.Path)
-               } else {
-                       // we already used the $PATH, so no fallback
+       defer func() {
+               if err != nil {
                        if err := os.Remove(partialTargetPath); err != nil {
-                               logger.Noticef("after encountering error applying delta: %v, also failed to remove partial delta target %q: %s", runErr, partialTargetPath, err)
+                               logger.Noticef("cannot remove partial file: %v", err)
                        }
-                       return err
+               }
+       }()
+
+       // try xdelta from the snap first
+       args := []string{"-d", "-s", snapPath, deltaPath, partialTargetPath}
+       cmd, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3", args...)
+       if err == nil {
+               err = cmd.Run()
+       }
+       if err != nil {
+               // try system xdelta next
+               logger.Noticef("cannot use internal xdelta: %v", err)
+               cmd = exec.Command("xdelta3", args...)
+               if err = cmd.Run(); err != nil {
+                       return fmt.Errorf("cannot use system xdelta: %v", err)
                }
        }
```
```

Reply:
thanks for the suggestion, but as per Pawel and Samuele's comment I had to go with a entirely different implementation